File: src/python/xerxes/core/prompt_template.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Ordered prompt sections used by the legacy template builder.

Defines :class:`PromptSection` (an enum of the canonical pieces a system
prompt can have) and :class:`PromptTemplate` (a dataclass holding the
per-section header labels and the rendering order). The newer streaming
loop builds prompts directly, but legacy modules still emit them through
this template for parity.
"""

from dataclasses import dataclass
from enum import Enum

SEP = "  "


class PromptSection(Enum):
    """Canonical section names used to slot content into a prompt template."""

    SYSTEM = "system"
    PERSONA = "persona"
    RULES = "rules"
    FUNCTIONS = "functions"
    TOOLS = "tools"
    EXAMPLES = "examples"
    CONTEXT = "context"
    HISTORY = "history"
    PROMPT = "prompt"


@dataclass
class PromptTemplate:
    """Per-section header labels plus the order they should be rendered.

    Attributes:
        sections: Map from :class:`PromptSection` to its header text.
        section_order: Order to render sections in the final prompt.
    """

    sections: dict[PromptSection, str] | None = None
    section_order: list[PromptSection] | None = None

    def __post_init__(self):
        """Populate ``sections`` and ``section_order`` with sensible defaults if missing."""
        self.sections = self.sections or {
            PromptSection.SYSTEM: "SYSTEM:",
            PromptSection.RULES: "RULES:",
            PromptSection.FUNCTIONS: "FUNCTIONS:",
            PromptSection.TOOLS: f"TOOLS:\n{SEP}When using tools, follow this format:",
            PromptSection.EXAMPLES: f"EXAMPLES:\n{SEP}",
            PromptSection.CONTEXT: "CONTEXT:\n",
            PromptSection.HISTORY: f"HISTORY:\n{SEP}Conversation so far:\n",
            PromptSection.PROMPT: "PROMPT:\n",
        }

        self.section_order = self.section_order or [
            PromptSection.SYSTEM,
            PromptSection.RULES,
            PromptSection.FUNCTIONS,
            PromptSection.TOOLS,
            PromptSection.EXAMPLES,
            PromptSection.CONTEXT,
            PromptSection.HISTORY,
            PromptSection.PROMPT,
        ]

File: xerxes/src/core/promptTemplate.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

export const SEP = '  '

export const PromptSection = {
  SYSTEM: 'system',
  PERSONA: 'persona',
  RULES: 'rules',
  FUNCTIONS: 'functions',
  TOOLS: 'tools',
  EXAMPLES: 'examples',
  CONTEXT: 'context',
  HISTORY: 'history',
  PROMPT: 'prompt',
} as const

export type PromptSection = (typeof PromptSection)[keyof typeof PromptSection]

export type PromptSections = Readonly<Partial<Record<PromptSection, string>>>

export const DEFAULT_PROMPT_SECTIONS: Readonly<Record<PromptSection, string>> = Object.freeze({
  [PromptSection.SYSTEM]: 'SYSTEM:',
  [PromptSection.PERSONA]: 'PERSONA:',
  [PromptSection.RULES]: 'RULES:',
  [PromptSection.FUNCTIONS]: 'FUNCTIONS:',
  [PromptSection.TOOLS]: 'TOOLS:\n' + SEP + 'When using tools, follow this format:',
  [PromptSection.EXAMPLES]: 'EXAMPLES:\n' + SEP,
  [PromptSection.CONTEXT]: 'CONTEXT:\n',
  [PromptSection.HISTORY]: 'HISTORY:\n' + SEP + 'Conversation so far:\n',
  [PromptSection.PROMPT]: 'PROMPT:\n',
})

export const DEFAULT_PROMPT_SECTION_ORDER: readonly PromptSection[] = Object.freeze([
  PromptSection.SYSTEM,
  PromptSection.RULES,
  PromptSection.FUNCTIONS,
  PromptSection.TOOLS,
  PromptSection.EXAMPLES,
  PromptSection.CONTEXT,
  PromptSection.HISTORY,
  PromptSection.PROMPT,
])

export interface PromptTemplateOptions {
  readonly sectionOrder?: readonly PromptSection[]
  readonly sections?: PromptSections
}

/**
 * Ordered prompt-section metadata shared by legacy prompt assembly code.
 *
 * The streaming runtime normally passes a complete system prompt directly,
 * while callers that need structured assembly can render one deterministic
 * prompt through this portable representation.
 */
export class PromptTemplate {
  readonly sectionOrder: readonly PromptSection[]
  readonly sections: Readonly<Record<PromptSection, string>>

  constructor(options: PromptTemplateOptions = {}) {
    this.sections = Object.freeze({
      ...DEFAULT_PROMPT_SECTIONS,
      ...options.sections,
    })
    this.sectionOrder = Object.freeze([...(options.sectionOrder ?? DEFAULT_PROMPT_SECTION_ORDER)])
  }

  /** Render only supplied non-blank sections in their configured order. */
  render(values: PromptSections, separator = '\n\n'): string {
    const parts: string[] = []
    for (const section of this.sectionOrder) {
      const content = values[section]?.trim()
      if (!content) continue
      const header = this.sections[section]?.trim()
      parts.push(header ? header + '\n' + content : content)
    }
    return parts.join(separator)
  }
}
